talk/expert/cpp20spaceship.tex

```diff
@@ -0,0 +1,242 @@
+\subsection[concepts]{The three-way comparison operator}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Motivation}
+    \begin{itemize}
+    \item There are frequent situations where one need some {\it <} operator for a home-made class. Typically, if you want to make a {\it std::vector} of such objects, and sort it. But also if you want to use it as a key for some {\it std::set} or {\it std::map}.
+    \item For completeness, one will also add {\it >}, {\it >=}, and {\it <=}, implemented as a reuse of either {\it <} and {\it ==}, or {\it <} and {\it >}.
+    \item Those operators will be defined as an external functions, optionally friends, so that left and right arguments will be similarly convertibles.
+    \item Much boilerplate code to write. Too much..
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Idea}
+    \begin{itemize}
+    \item C++20 introduce the new \textbf{spaceship} operator : {\it <=>}. Well, the real official name is \textbf{three-way comparison operator}.
+    \item It is provided by default for all predefined types, and return \textbf{something} which can be compared to {\it 0} (such as the {\it std::strcmp}), meaning \textbf{lower than} if this \textbf{something} is lower than {\it 0}, \textbf{greater than} if it is greater than {\it 0}, and \textbf{equivalent} if it is equal {\it 0}:
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    \begin{cppcode*}{}
+#include <iostream>
+
+template <typename T>
+void three_way_compare( T lhs, T rhs )
+ {
+  auto result = (lhs<=>rhs) ;
+  std::cout
+    <<lhs<<" <=> "<<rhs<<": "<<'<'
+    <<(result<0)<<'|'<<(result==0)<<'|'<<(result>0)
+    <<'>'<<std::endl ;
+ }
+
+int main()
+ {
+  three_way_compare(1,2) ;
+  three_way_compare(2,2) ;
+  three_way_compare(2,1) ;
+ }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{The three-way comparison operator}
+  \begin{block}{Different kinds of ordering}
+    \begin{itemize}
+    \item The real return type of {\it <=>} for integers is {\it std::strong_ordering} : whatever the values, you will always get {\it true} for exactly one test among {\it <0}, {\it ==0}, and {\it >0}.
+    \item On the contrary, the return type of {\it <=>} for floating point numbers is {\it std::partial_ordering}, because sometimes the tree tests may return {\it false}, typically if one one the number is {\it NaN}.
```

Comment:
Probably NaN has not been adressed yet. Since we never talk about floating point gory details. Yet, we should have some example which justify the existence of such partial_ordering... Or we keep this section a lot simpler, not talking about the different kinds of ordering ?

Reply:
Hmm, leave it. `float` is a good example of a partially ordered type.